### PR TITLE
db/hints: Clean up hint_storage.cc

### PR DESCRIPTION
**Pull request by dawmd**
This PR is the second step in refactoring the Hinted Handoff module. It cleans up the contents of the file `hint_storage.cc`. The biggest change is the transition from continuations to coroutines.

Refs #15358

### CODE REVIEW

**Comment by kbr-scylla** [comment]
@dawmd please request relevant people to review

**Comment by dawmd** [comment]
The failed tests seem to be flaky. I ran them locally again and everything passes; they also seem unrelated to my changes.

File: db/hints/internal/hint_storage.cc [hunk 1]
```diff
@@ -50,16 +50,24 @@ using hints_ep_segments_map = std::unordered_map<unsigned, std::list<fs::path>>;
...
         try {
-            shard_id = std::stoi(de.name.c_str());
+            shard_id = std::stoi(de.name);
         } catch (std::invalid_argument& ex) {
             manager_logger.debug("Ignore invalid directory {}", de.name);
             return make_ready_future<>();
         }
-        return f(std::move(dir), std::move(de), shard_id);
+
+        return futurize_invoke(std::forward<Func>(func), std::move(dir), std::move(de), shard_id);
```
Comment (piodul) [inline db/hints/internal/hint_storage.cc:70]: Is `futurize_invoke` here necessary?

File: db/hints/internal/hint_storage.cc [hunk 2]
```diff
@@ -50,16 +50,24 @@ using hints_ep_segments_map = std::unordered_map<unsigned, std::list<fs::path>>;
 // map: IP -> map: shard -> segments
 using hints_segments_map = std::unordered_map<sstring, hints_ep_segments_map>;
 
-future<> scan_for_hints_dirs(const fs::path& hints_directory, std::function<future<> (fs::path dir, directory_entry de, unsigned shard_id)> f) {
```
Comment (piodul) [inline db/hints/internal/hint_storage.cc]: An advantage of using std::function here is that you can give names to its arguments. This is lost when using concepts.

File: db/hints/internal/hint_storage.cc [hunk 3]
```diff
@@ -50,16 +50,24 @@ using hints_ep_segments_map = std::unordered_map<unsigned, std::list<fs::path>>;
...
-    return lister::scan_dir(hints_directory, lister::dir_entry_types::of<directory_entry_type::directory>(), [f = std::move(f)] (fs::path dir, directory_entry de) mutable {
+/// The caller must ensure that the passed functor's captures
+/// are kept alive at least until the returned future resolves.
+/// It also applies to the functor itself if it's passed as an lvalue reference.
+template <typename Func>
+    requires std::invocable<Func, fs::path, directory_entry, unsigned>
+future<> scan_for_hints_dirs(const fs::path& hint_directory, Func&& func) {
```
Comment (piodul) [inline db/hints/internal/hint_storage.cc]: The function is not on a hot path so I'm not sure it's worth using templates/concepts instead of just std::function.

File: db/hints/internal/hint_storage.cc [hunk 4]
```diff
@@ -71,34 +71,42 @@ future<> scan_shard_hint_directories(const fs::path& hint_directory, Func&& func
     });
 }
 
-/// \brief Scan the given hints directory and build the map of all present hints segments.
+/// \brief Scan the given hint directory and build the map of all present hint segments.
```
Comment (piodul) [inline db/hints/internal/hint_storage.cc:74]: > db/hints: Clean up get_current_hints_segments in hint_storage.cc 

Please describe briefly in the commit message what those cleanups are about.

File: db/hints/internal/hint_storage.cc [hunk 5]
```diff
@@ -109,58 +109,61 @@ hints_segments_map get_current_hints_segments(const fs::path& hint_directory) {
     return current_hint_segments;
 }
 
-/// \brief Rebalance hints segments for a given (destination) end point
+/// \brief Rebalance hint segments for a given (destination) end point
```
Comment (piodul) [inline db/hints/internal/hint_storage.cc:110]: > db/hints: Clean up rebalance_segments_for in hint_storage.cc 

Ditto

File: db/hints/internal/hint_storage.cc [hunk 6]
```diff
@@ -168,71 +168,82 @@ void rebalance_segments_for(const sstring& ep, size_t segments_per_shard,
     }
 }
 
-/// \brief Rebalance all present hints segments.
+/// \brief Rebalance all present hint segments.
```
Comment (piodul) [inline db/hints/internal/hint_storage.cc:167]: >  db/hints: Clean up rebalance_segments in hint_storage.cc

Ditto

File: db/hints/internal/hint_storage.cc [hunk 7]
```diff
@@ -168,71 +168,82 @@ void rebalance_segments_for(const sstring& ep, size_t segments_per_shard,
...
         if (q) {
-            rebalance_segments_for(ep, q, hints_directory, current_segments_map, current_segments_to_move);
+            rebalance_segments_for(ep, q, hint_directory, current_segments_map,
+                    current_segments_to_move);
         }
 
         if (r) {
-            rebalance_segments_for(ep, q + 1, hints_directory, current_segments_map, current_segments_to_move);
+            rebalance_segments_for(ep, q + 1, hint_directory, current_segments_map,
+                    current_segments_to_move);
```
Comment (piodul) [inline db/hints/internal/hint_storage.cc]: Nit: the seastar coding style requires lines to be at most 120 characters in length. Lines shorter than that limit are fine, so no need to change them - for example, these ones don't seem to need such an adjustment.

File: db/hints/internal/hint_storage.cc [hunk 8]
```diff
@@ -248,23 +248,25 @@ void rebalance_segments(const fs::path& hint_directory, hints_segments_map& segm
     }
 }
 
-/// \brief Remove sub-directories of shards that are not relevant any more (re-sharding to a lower number of shards case).
+/// \brief Remove subdirectories of shards that are not relevant anymore (re-sharding to a lower number of shards case).
```
Comment (piodul) [inline db/hints/internal/hint_storage.cc:245]: > db/hints: Clean up remove_irrelevant_shards_directories 

... and so on ...

File: db/hints/internal/hint_storage.cc (hunk 1 above)
Comment (dawmd) [reply db/hints/internal/hint_storage.cc:70]: Yes. `lister::scan_dir` requires that the passed functor return `future<>`, but it might make sense to pass a void function to the function too. `futurize_invoke` is a way to ensure those functors are also valid arguments.

File: db/hints/internal/hint_storage.cc (hunk 2 above)
Comment (dawmd) [reply db/hints/internal/hint_storage.cc]: This can be "simulated" using comments:
```cc
std::invocable<Func, fs::path /* hint_directory */, directory_entry, unsigned /* shard_id */>
```

File: db/hints/internal/hint_storage.cc (hunk 3 above)
Comment (dawmd) [reply db/hints/internal/hint_storage.cc]: You're correct. I was also wondering if moving to templates might be an overkill here if we end up using I/O anyway. There were two reasons why I decided to go with this solution:

* allowing for functors returning `void`. This could be achieved using `std::function` (or `noncopyable_function`/`std::move_only_function`) too, but we'd end up with a template then anyway,

* though the gain is extremely small, this approach also saves some memory and potential allocations, as well as prevents virtual dispatch. Nowadays, all major compilers use small object optimisation in `std::function` so that it doesn't allocate memory dynamically if the functor's size isn't too big. However, if, for example, the functor has no captures at all, then instead of taking no space (as it could be optimised away by the compiler) or just one byte, we end up with something bigger. The virtual dispatch (being an inherent part of `std::function`) isn't a good thing either as it's more likely that we end up with a cache miss + we need one more level of indirection. Of course, `utils::scan_dir` takes an `std::function` as its argument anyway, so these things will happen, but since we're effectively passing the lambda created in this function's body and the passed functor is only captured by it, we can at least optimise that.

The latter argument isn't very important as I/O will dominate the time this function takes to execute anyway. I mostly had in mind that in the future we might want to reuse the function in other contexts where it could matter more (though probably not).

The first argument might be relevant, but on the other hand, the caller could just wrap their functor in a lambda returning `future<>` and that's it. At the same time, these changes introduce more mental effort on the programmer because they need to ensure that the functor will live at least until the returned future resolves, while with `std::function` we have it, at least in some cases, for free.

I can go back to `std::function`.

**Comment by dawmd** [comment]
V3:
* moved back to `std::function`
* elaborated on the changes in commit messages

File: db/hints/internal/hint_storage.cc [hunk 9]
```diff
@@ -74,16 +73,14 @@ future<> scan_shard_hint_directories(const fs::path& hint_directory,
...
 /// \return a map: ep -> map: shard -> segments (full paths)
-hints_segments_map get_current_hints_segments(const fs::path& hint_directory) {
+future<hints_segments_map> get_current_hints_segments(const fs::path& hint_directory) {
     hints_segments_map current_hint_segments{};
 
     // Shard level.
-    scan_shard_hint_directories(hint_directory,
-            [&current_hint_segments] (fs::path dir, directory_entry de, unsigned shard_id) {
+    co_await scan_shard_hint_directories(hint_directory,
+            coroutine::lambda([&current_hint_segments] (fs::path dir, directory_entry de, unsigned shard_id) {
```
Comment (piodul) [inline db/hints/internal/hint_storage.cc:84]: You are using `coroutine::lambda` here as an argument to `scan_shard_hint_directories`, but not in `remove_irrelevant_shards_directories`. I don't think it is needed here as `lister::scan_dir` used in `scan_shard_hint_directories` will keep the std::function passed to it alive until it finishes its work.

File: db/hints/internal/hint_storage.cc [hunk 10]
```diff
@@ -74,16 +73,14 @@ future<> scan_shard_hint_directories(const fs::path& hint_directory,
...
 /// \param hint_directory directory to scan
 /// \return a map: ep -> map: shard -> segments (full paths)
-hints_segments_map get_current_hints_segments(const fs::path& hint_directory) {
+future<hints_segments_map> get_current_hints_segments(const fs::path& hint_directory) {
     hints_segments_map current_hint_segments{};
 
     // Shard level.
-    scan_shard_hint_directories(hint_directory,
-            [&current_hint_segments] (fs::path dir, directory_entry de, unsigned shard_id) {
+    co_await scan_shard_hint_directories(hint_directory,
```
Comment (piodul) [inline db/hints/internal/hint_storage.cc:83]: > db/hints: Move to coroutines in hint_storage.cc 

Nit: "Move to" at the beginning of the commit message strongly suggests that you are moving code from one place to another. We usually use the term "coroutinize" to indicate that some code was adjusted to use coroutines.

File: db/hints/internal/hint_storage.cc [hunk 11]
```diff
@@ -249,36 +242,34 @@ void rebalance_segments(const fs::path& hint_directory, hints_segments_map& segm
...
-        rebalance_segments(hints_directory, current_hints_segments);
+    // Move segments to achieve an even distribution of files among all present shards.
+    co_await rebalance_segments(hints_directory, current_hints_segments);
 
-        // Remove the directories of shards that are not present anymore - they should not have any segments by now
-        remove_irrelevant_shards_directories(hints_directory);
-    });
+    // Remove the directories of shards that are not present anymore.
+    // They should not have any segments by now.
+    co_await remove_irrelevant_shards_directories(hints_directory);
```
Comment (piodul) [inline db/hints/internal/hint_storage.cc]: Nit: indentation changes should usually be done in a separate commit, to improve the readability of diffs.

File: db/hints/internal/hint_storage.cc [hunk 12]
```diff
@@ -43,12 +43,14 @@ namespace fs = std::filesystem;
 namespace db::hints {
 namespace internal {
 
+namespace {
+
```
Comment (piodul) [inline db/hints/internal/hint_storage.cc:46]: > An anonymous namespace should be preferred wherever possible.

Why?

File: db/hints/internal/hint_storage.cc (hunk 9 above)
Comment (dawmd) [reply db/hints/internal/hint_storage.cc:84]: > You are using `coroutine::lambda` here as an argument to `scan_shard_hint_directories`, but not in `remove_irrelevant_shards_directories`.

That's because the lambda used in the latter function doesn't have any captures, so there's no need to wrap it in a `seastar::coroutine::lambda`.

> I don't think it is needed here as `lister::scan_dir` used in `scan_shard_hint_directories` will keep the `std::function` passed to it alive until it finishes its work.

But this is a very good point. I'll change it.

File: db/hints/internal/hint_storage.cc (hunk 12 above)
Comment (dawmd) [reply db/hints/internal/hint_storage.cc:46]: That's a valid question. `static` functions and those wrapped in an anonymous namespace are equivalent ways to achieve the same, so these changes don't actually change anything. The only explanation I can provide is:
* I personally always prefer anonymous namespaces except for some niche cases, so this is my coding style preference (which is not relevant for non-personal projects of course)
* when adding new things, it's easy to forget about the `static`. But if you put something in an anonymous namespace that shouldn't be there, the linker will immediately start screaming (in most cases) because some symbol will be undefined. This solution seems less error-prone to me as a result.

If you think this commit should be skipped, I can drop it. But if it's fine and just the commit message should be better (as in provide some relevant justification of the changes), I can try to reword the second bullet from this message.

Thank you for stressing on this. I should think whether changes I make are really something that should be done or perhaps not.

Comment (piodul) [reply db/hints/internal/hint_storage.cc:46]: That's fine, I just wanted to know the reason. Clear descriptions in commits are also important because they convey your intent to reviewers and then they can verify whether what you wrote aligns with it or not.

I don't think our coding style disallows anonymous namespaces or encourages them (I couldn't find anything in coding-style.md), but I can see that we do use them in the code, so it's fine.

**Comment by dawmd** [comment]
V4:
* Got rid of the unnecessary `seastar::coroutine::lambda`.
* Reworded the commit message related to the anonymous namespace.
* Reworded the commit message related to coroutinizing the file.
* Split a commit to fix indentation in another one and make reviewing code easier.

**Review by piodul — APPROVED** [review]

**Comment by dawmd** [comment]
@scylladb/scylla-maint could you take a look at this PR?

**Comment by denesb** [comment]
I will queue this as-is but for the future please note that excessive style-only changes are usually frowned upon because such changes are disruptive and style is also subjective to a certain degree so there is a possibility to produce infinite amount of style-only patches. For this reason we usually discourage them. As a developer, you will have to learn to work with code which is not well formatted according to your taste, or even according to coding guidelines that should have been enforced on that code.

So as you continue refactoring the hint code, please keep this in mind and reduce the amount of style-only cleanups.
